File: solutions/python/dnd-character/2/dnd_character.py
import secrets

ABILITIES = (
    "strength",
    "dexterity",
    "constitution",
    "intelligence",
    "wisdom",
    "charisma",
)


def modifier(score: int) -> int:
    """Calculate the modifier for a given ability score.

    Args:
        score (int): The ability score.

    Returns:
        int: The calculated modifier.
    """
    return (score - 10) // 2


class Character:
    """Represents a game character with various abilities and hit points."""

    def __init__(self):
        """Initialize a new character with random ability scores and hit points."""
        for ability in ABILITIES:
            setattr(self, ability, self.ability())
        self.hitpoints = 10 + modifier(self.constitution)

    @staticmethod
    def ability() -> int:
        """Roll four six-sided dice and return the sum of the three highest rolls.

        Returns:
            int: The ability score.
        """
        dices = sorted(secrets.randbelow(7) for _ in range(4))
        return sum(dices[1:])

File: solutions/lua/dnd-character/2/dnd-character.lua
local function roll_dice()
  return {
    math.random(1, 6),
    math.random(1, 6),
    math.random(1, 6),
    math.random(1, 6)
  }
end

local function ability()
  local scores = roll_dice()
  table.sort(scores, function(a, b) return a > b end)
  return scores[3] + scores[2] + scores[1]
end

local function modifier(input)
  return math.floor((input - 10) / 2)
end

local function new_character(self, name)
  local constitution = ability()

  return {
    name = name,
    strength = ability(),
    dexterity = ability(),
    constitution = constitution,
    intelligence = ability(),
    wisdom = ability(),
    charisma = ability(),
    hitpoints = 10 + modifier(constitution)
  }
end

return {
  Character = { new = new_character },
  ability = ability,
  roll_dice = roll_dice,
  modifier = modifier
}